File: 0198-house-robber/0198-house-robber.py
class Solution:
    def solve(self, nums, store, n, i):
        if i >= n:
            return 0

        if store[i] != -1 : return store[i]

        take = nums[i] + self.solve(nums, store, n, i+2)
        notTake = self.solve(nums, store, n, i+1)

        store[i] = max(take, notTake)

        return store[i]

    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        store = [-1] * (n+1)
        store[n] = 0
        store[n-1] = nums[n-1]
        for i in range(n-2, -1, -1):
            store[i] = max(store[i+1], store[i+2] + nums[i])
        return store[0]

File: 0198-house-robber/0198-house-robber.java
class Solution {
    static int[] nums;
    int[] store;

    public int solve(int i){
        if (i >= nums.length) return 0;
        if (store[i] != -1) return store[i];

        int take = nums[i] + solve(i+2);
        int notTake = solve(i+1);

        return store[i] = Math.max(take, notTake);
    }

    public int rob(int[] nums) {
        this.nums = nums;
        store = new int[nums.length];
        java.util.Arrays.fill(store, -1);
        return solve(0);
    }
}